rust/processor/src/db_writer.rs

```diff
@@ -0,0 +1,325 @@
+        Item: field_count::FieldCount
+            + serde::Serialize
+            + for<'de> serde::Deserialize<'de>
+            + Send
+            + Sync
+            + Clone
+            + 'static,
+        Query:
+            QueryFragment<diesel::pg::Pg> + diesel::query_builder::QueryId + Send + Sync + 'static,
+    {
+        // TODO: ideally this kind of skip is done much earlier in the process
+        if self.skip_tables.contains(table_name) {
+            return;
+        }
+
+        DB_EXECUTOR_CHANNEL_CHUNK_INSERT_COUNT
+            .with_label_values(&[self.processor_name, table_name])
+            .inc();
+
+        let chunk_size = self.chunk_size::<Item>(table_name);
+        let chunks = get_chunks(items_to_insert.len(), chunk_size);
+        for (start_ind, end_ind) in chunks {
+            let items = items_to_insert[start_ind..end_ind].to_vec();
+            let data_with_query = DataWithQuery::new(items, query_fn);
+
+            self.query_sender
+                .send(QueryGenerator {
+                    table_name,
+                    db_executable: Box::new(data_with_query),
+                })
+                .await
+                .expect("Error sending query generator to db writer task");
+        }
+
+        DB_EXECUTOR_CHANNEL_SIZE
+            .with_label_values(&[self.processor_name])
+            .set(self.query_sender.len() as i64);
+    }
+}
+
+pub async fn launch_db_writer_tasks(
+    query_receiver: AsyncReceiver<QueryGenerator>,
+    processor_name: &'static str,
+    num_tasks: usize,
+    db_pool: PgDbPool,
+) {
+    info!(
+        processor_name = processor_name,
+        service_type = PROCESSOR_SERVICE_TYPE,
+        "[Parser] Starting db writer tasks",
+    );
+    let tasks = (0..num_tasks)
+        .map(|_| launch_db_writer_task(query_receiver.clone(), processor_name, db_pool.clone()))
+        .collect::<Vec<_>>();
+    futures::future::try_join_all(tasks)
+        .await
+        .expect("Error joining db writer tasks");
+}
+
+// This is a wrapper struct that allows us to pass additional data alongside the DbExecutable
+pub struct QueryGenerator {
+    pub table_name: &'static str,
+    pub db_executable: Box<dyn DbExecutable>,
+}
+
+pub fn launch_db_writer_task(
+    query_receiver: AsyncReceiver<QueryGenerator>,
+    processor_name: &'static str,
+    db_pool: PgDbPool,
+) -> JoinHandle<()> {
+    tokio::spawn(async move {
+        loop {
+            match query_receiver.recv().await {
+                Ok(mut query_generator) => {
+                    let query_res = query_generator
+                        .db_executable
+                        .execute_or_retry_cleaned(
+                            processor_name,
+                            query_generator.table_name,
+                            db_pool.clone(),
+                        )
+                        .await;
+                    query_res.expect("Error executing query");
+                    drop(query_generator);
```

Comment:
do we need to explicitly drop here?
should we drop it in a background thread?

Reply:
I explicitly drop so that I can be guaranteed it holds across this unsafe block